### PR TITLE
Fixes next vitis

### PR DESCRIPTION
**Pull request by Ralender**
This patch fixes some of the issues under the next version of vitis. also adds xrt_alloc which should be used when using xrt directly or not.

### CODE REVIEW

**Comment by keryell** [comment]
I guess you need to rebase it on latest `sycl/unified/next` I have fixed.

**Comment by Ralender** [comment]
here is the clean rebase

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp [hunk 1]
```diff
@@ -282,6 +287,7 @@ struct LSMDState {
     std::vector<Value *> Args;
     for (auto &A : F->args())
       Args.push_back(&A);
+    std::ignore = std::initializer_list<int>{(Args.push_back(ts), 0)...};
```
Comment (keryell) [inline llvm/lib/SYCL/LowerSYCLMetaData.cpp]: Why do you need this here but not on the previous line too?

File: sycl/include/sycl/ext/xilinx/xrt_alloc.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+//==--------- xrt_alloc.hpp - SYCL XRT Allocator ---------------------------==//
+//
```
Comment (keryell) [inline sycl/include/sycl/ext/xilinx/xrt_alloc.hpp:2]: Explain somewhere why you need this.

File: sycl/test/vitis/disabled/pipe_interprocedural.cpp [hunk 3]
```diff
@@ -9,6 +9,8 @@
 #include <sycl/sycl.hpp>
 #include <sycl/ext/intel/fpga_extensions.hpp>
 
+using namespace sycl;
```
Comment (keryell) [inline sycl/test/vitis/disabled/pipe_interprocedural.cpp:12]: I think you should not need this.

File: llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp [hunk 4]
```diff
@@ -174,6 +174,10 @@ static cl::opt<unsigned>
                            cl::desc("Default threshold (max size of unrolled "
                                     "loop), used in all but O3 optimizations"));
 
+static cl::opt<bool> UnrollOnlyWhenForced(
+  "unroll-only-when-forced", cl::Hidden,
+  cl::desc("Allow the loop remainder to be unrolled."));
```
Comment (keryell) [inline llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp]: Explain the rationale

File: llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp [hunk 5]
```diff
@@ -34,7 +38,21 @@ Value *Skip1BitCast(Value *V) {
   return V;
 }
 
-/// Choose the LLVM type that should be used for a memory operation on Ptr1 and maybe Ptr2
+void annotateBackEdge(BranchInst *Br) {
```
Comment (keryell) [inline llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp]: Not clear what this does, considered the name of the function.

**Review by keryell — CHANGES_REQUESTED** [review]
You should clarify some points.

**Comment by Ralender** [comment]
The review has not yet been addressed.

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp (hunk 1 above)
Comment (Ralender) [reply llvm/lib/SYCL/LowerSYCLMetaData.cpp]: Why would I need what on the previous line ?

File: llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp (hunk 4 above)
Comment (Ralender) [reply llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp]: Hope the updated description helps

File: llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp (hunk 5 above)
Comment (Ralender) [reply llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp]: is the new name more clear ?

**Review by Ralender — COMMENTED** [review]
I added a commit to deal with terminate_xsim

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 6]
```diff
@@ -1005,6 +1005,9 @@ static bool UpgradeIntrinsicFunction1(Function *F, Function *&NewFn) {
       return true;
   }
 
+  if (F->getIntrinsicID() == Intrinsic::not_intrinsic)
```
Comment (keryell) [inline llvm/lib/IR/AutoUpgrade.cpp:1012]: Why it was not the case before?
+ Comment.

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp [hunk 7]
```diff
@@ -282,6 +287,7 @@ struct LSMDState {
     std::vector<Value *> Args;
     for (auto &A : F->args())
       Args.push_back(&A);
+    (void)std::initializer_list<int>{(Args.push_back(ts), 0)...};
```
Comment (keryell) [inline llvm/lib/SYCL/LowerSYCLMetaData.cpp:290]: No more `std::ignore =`?

File: llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp (hunk 5 above)
Comment (keryell) [reply llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp]: Yes.

**Review by keryell — COMMENTED** [review]
There are still some unanswered comments.

File: llvm/lib/SYCL/LowerSYCLMetaData.cpp (hunk 7 above)
Comment (Ralender) [reply llvm/lib/SYCL/LowerSYCLMetaData.cpp:290]: I prefer (void)

File: sycl/test/vitis/disabled/pipe_interprocedural.cpp (hunk 3 above)
Comment (Ralender) [reply sycl/test/vitis/disabled/pipe_interprocedural.cpp:12]: This is the way it is for most tests.

File: sycl/include/sycl/ext/xilinx/xrt_alloc.hpp [hunk 8]
```diff
@@ -5,6 +5,12 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
+//
+// When using Xilinx/AMD FPGA through the OpenCL backend. Iff buffer allocation
+// are not aligned on 4096 bytes, XRT will perform an extra memcpy. This is file
+// provides a sycl allocator with the proper alignment.
```
Comment (keryell) [inline sycl/include/sycl/ext/xilinx/xrt_alloc.hpp]:
```suggestion
// When using Xilinx/AMD FPGA through the OpenCL backend, if buffer allocation
// are not aligned on 4096 bytes, XRT will perform an extra memcpy. This file
// provides a SYCL allocator with the proper alignment.
```

**Review by keryell — CHANGES_REQUESTED** [review]
Almost there!

**Review by keryell — APPROVED** [review]
Thanks!

**Pull request merged**